### PR TITLE
CI: Add TODO/FIXME checker

### LINKED ISSUE #444: Add TODO/FIXME CI test
`TODO` and `FIXME` comments in the code tend to sit there and not get fixed. Much better that rather than such comments we have a reference to an issue that explains the problem properly.

An issue provides greater visibility thus maximising the chance the problem will be resolved.

Since we already screen for such comments in manual code review, why not automate it?

The plan is to fail a PR if it contains such a comment **iff** there is no associated issue URL with it.

Like other similar CI checks, this will be added to https://github.com/kata-containers/tests/blob/master/.ci/static-checks.sh and will only fail the CI run if the user modifies a file containing such a comment. If the CI is run on `master`, any such comments will cause a failure.

### PR DESCRIPTION
**Pull request by jodh-intel**
Alert users to TODO/FIXME comments that tend never to get fixed by
failing the CI *unless* those comments also refer to a github issue
where the problem is being tracked.

Fixes #444.

Signed-off-by: James O. D. Hunt <james.o.hunt@intel.com>

### CODE REVIEW

File: .ci/static-checks.sh [hunk 1]
```diff
@@ -429,6 +429,71 @@ check_docs()
 	rm -f "$url_map" "$invalid_urls"
 }
 
+# Tests to apply to all files.
+#
+# Currently just looks for TODO/FIXME comments that should be removed.
```
Comment (grahamwhaley) [inline .ci/static-checks.sh]: s/removed/converted to or annotated with an Issue/ ?

**Comment by grahamwhaley** [comment]
Battoning down the hatches eh? I can't dis-agree, but am bracing myself for the flurry of failed PRs that touch a file that already has an errant comment in it...
Out of interest then, have you run this on the current code bases, and what did you find - do we have many current offending comments?

File: .ci/static-checks.sh (hunk 1 above)
Comment (jodh-intel) [reply .ci/static-checks.sh]: Good call - fixed.

**Comment by jodh-intel** [comment]
Well, yes. But we need to collectively agree this is a good idea so I'll step back and let folk ack/nack :)

We're actually in pretty good shape. Calling just that new function on all the kata repos from a script that just adds in the name of the repo it's checking, we have the following:

```
INFO: Checking repo https://github.com/kata-containers/agent
INFO: Checking files
ERROR: The following files contain TODO/FIXME's that need converting to issues:

./grpc_test.go
./Dockerfile
./protocols/grpc/agent.proto
./protocols/grpc/agent.pb.go
./grpc.go

INFO: Checking repo https://github.com/kata-containers/ci
INFO: Checking files
ERROR: The following files contain TODO/FIXME's that need converting to issues:

./VMs/metrics/3_complete_baseimage.sh
./VMs/metrics/global_env.sh

INFO: Checking repo https://github.com/kata-containers/community
INFO: Checking files
INFO: Ignoring comment tag in document ./PR-Review-Guide.md
INFO: Checking repo https://github.com/kata-containers/documentation
INFO: Checking files
INFO: Checking repo https://github.com/kata-containers/ksm-throttler
INFO: Checking files
INFO: Checking repo https://github.com/kata-containers/osbuilder
INFO: Checking files
INFO: Checking repo https://github.com/kata-containers/packaging
INFO: Checking files
ERROR: The following files contain TODO/FIXME's that need converting to issues:

./scripts/configure-hypervisor.sh
./.ci/teardown.sh
./.ci/run.sh

INFO: Checking repo https://github.com/kata-containers/proxy
INFO: Checking files
INFO: Checking repo https://github.com/kata-containers/runtime
INFO: Checking files
ERROR: The following files contain TODO/FIXME's that need converting to issues:

./virtcontainers/container.go
./virtcontainers/qemu.go
./virtcontainers/sandbox.go
./virtcontainers/hyperstart_agent.go
./virtcontainers/kata_builtin_proxy.go
./virtcontainers/proxy.go
./virtcontainers/qemu_arch_base.go
./virtcontainers/network.go
./virtcontainers/kata_agent.go
./virtcontainers/filesystem.go
./virtcontainers/nsenter.go

INFO: Checking repo https://github.com/kata-containers/shim
INFO: Checking files
INFO: Checking repo https://github.com/kata-containers/tests
INFO: Checking files
ERROR: The following files contain TODO/FIXME's that need converting to issues:

./metrics/density/docker_memory_usage.sh
./cmd/checkmetrics/main.go
./cmd/checkmetrics/compare.go
./cmd/checkmetrics/compare_test.go

INFO: Checking repo https://github.com/kata-containers/tests.16may2018-before-force-push
INFO: Checking files
ERROR: The following files contain TODO/FIXME's that need converting to issues:

./metrics/density/docker_memory_usage.sh
./cmd/checkmetrics/main.go
./cmd/checkmetrics/compare.go
./cmd/checkmetrics/compare_test.go
```

But that's the total, so users will only need to fix those comments if they happen to change those files.

If we plan to land this, yes, we'll need to go round converting the files listed above though to ensure the `master` build doesn't bomb out ;)

**Comment by GabyCT** [comment]
lgtm

**Pull request merged**